### PR TITLE
[move-to-yul] Add support of arrays to dispatcher

### PR DESCRIPTION
**Pull request by rahxephon89**
## Motivation

Add marshaling of arrays (including bytes and string) to dispatcher.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Semantic tests are added to tests/test-dispatcher. More tests will be added as needed.

### CODE REVIEW

File: language/evm/move-to-yul/src/context.rs [hunk 1]
```diff
@@ -450,6 +450,15 @@ impl<'a> Context<'a> {
         }
     }
 
+    /// Returns true of the type allocates memory.
+    pub fn type_allocates_memory_not_vector(&self, ty: &Type) -> bool {
```
Comment (wrwg) [inline language/evm/move-to-yul/src/context.rs:454]: Can we use a better name for this? Like `type_is_proper_struct` or something.

File: language/evm/move-to-yul/src/dispatcher_generator.rs [hunk 2]
```diff
@@ -0,0 +1,1084 @@
+// Copyright (c) The Diem Core Contributors
```
Comment (wrwg) [inline language/evm/move-to-yul/src/dispatcher_generator.rs:1]: Love to see this in a new file!

**Review by wrwg — DISMISSED** [review]
Looks great. Only a nit, so feel free to merge and fix in the next PR.

File: language/evm/move-to-yul/src/context.rs (hunk 1 above)
Comment (rahxephon89) [reply language/evm/move-to-yul/src/context.rs:454]: Thanks! I will fix it in the next PR

**Pull request closed without merging**